### DISCUSSION TITLE
Gradual Fading Weight

### DISCUSSION THREAD

**Discussion by guzuligo** [Mikubill/sd-webui-controlnet#1071] [Ideas]
When I set the "Ending Control Steps" to a value, it sometimes causes a strange chaos at the time of the stop and change the result drastically.

I suggest adding a controller for a cooldown time where the weight is gradually reduced until it becomes zero. And the user can specify how many steps it takes till it becomes zero.

**Comment by marcsyp** [comment]
I like this idea just because it will allow for some interesting experiments.  That said, once you start implementing something like this, it will start to beg the questions -- should we also have a ramp up as well as ramp down?  Should you be able to set easing curves?  How do you handle start and stop points of easing with respect to the guidance values (which are in a 0-1.0 factor of the entire process as opposed to individual steps), and is there instead a syntax to defined your guidance weight over the timeline instead of a single crude control with additional parameters that have to added one by one.  Using number of steps is not ideal because it mixes two different declaration paradigms and would require constant manipulation when experimenting with number of steps for your image.

Consider this as a potential option:

A text box guidance override, with for parameters, all required (ease in start, peak start, peak end, ease out end) with easing options for in/out:
0, 0.2, 0.5, 0.7
Ease in: Linear (will ease in from 0-0.2)
Ease out: Start slow (will ease out slowly from 0.5-0.7)

0, 0.2, 0.5, 0.7
Ease in: None (no ControlNet influence from 0-0.2)
Ease out: Linear (ease out with linear decline from 0.5-0.7)

etc

**Reply by guzuligo** [reply]
I think a simple linear one is enough.

But, to give more control, you could just add a textbox that you can write simple python giving the user the time as an input, and user can write something like:

2.0 if (t<0.5) else 1.0

You can use an eval function to let that textbox be evaluated as a python code.

But, this might be better hidden under an "advanced" tab.

**Comment by ljleb** [comment]
Maybe add a textbox with custom syntax to specify dynamic weight behavior?

i.e. linear:
```
0-1.5
```

ease-out:
```
(0-1)-1
```

ease-in:
```
0.2-(0.2-1)
```

quadratic:
```
(0-1)-(1-0)
```

constant:
```
0.5
```

Note that this will be confusing without documentation and examples. However, it would make it possible to schedule fine-grain weights for users that know what they're doing. I have seen Mikubill mention this somewhere, IIRC it was something between 7 and 9 weights per model:
```
0,0-1,0.7,1,(0-1)-0,...
```

The syntax could be improved, if we want to consider this path. It is just an example used for illustrative purposes here.

For example, to stay consistent with the prompt syntax from the webui, we could use `[0-1 : 0.2]` or `[0-1 :: 0.8]` to only start considering weghts from and until a certain point, respectively. Maybe not the best idea, I am not sure.

**Comment by guzuligo** [comment]
I think the easiest way to implement is to have a textbox that uses eval() to python code. And use variable t for step, where it ranges from 0 to 1.
User can just write: 1-t and gets a fading weight.

If someone wants more control, he could just write an equation using python functions.
